File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* recur(ListNode* current, ListNode* previous){
        if(current==NULL)return previous;
        ListNode* forward = current->next;
        current->next = previous;
        return recur(forward,current);
    }

    ListNode* reverseList(ListNode* head) {
       
        return  recur(head, NULL);;
    }
};

File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode recur(ListNode current, ListNode previous){
        if(current==null)return previous;
        ListNode forward = current.next;
        current.next = previous;
        return recur(forward,current);
    }
    public ListNode reverseList(ListNode head) {
        return recur(head,null);
    }
}